crates/ra_hir/src/ty.rs

```diff
@@ -872,6 +873,155 @@ impl<'a, D: HirDatabase> InferenceContext<'a, D> {
+                    .variant_data
+                    .fields()
+                    .iter()
+                    .cloned()
+                    .collect();
+                Some((type_for_struct(self.db, s), fields))
+            }
+            Def::EnumVariant(ev) => {
+                let fields: Vec<_> = ev.variant_data(self.db).fields().iter().cloned().collect();
+                Some((type_for_enum_variant(self.db, ev), fields))
+            }
+            _ => None,
+        }
+    }
+
+    fn infer_tuple_struct(&mut self, path: Option<&Path>, subpats: &[PatId]) -> Ty {
+        let (ty, fields) = if let Some(x) = self.resolve_fields(path) {
+            x
+        } else {
+            return Ty::Unknown;
+        };
+
+        if fields.len() != subpats.len() {
+            return Ty::Unknown;
+        }
+
+        for (&subpat, field) in subpats.iter().zip(fields.iter()) {
+            let sub_ty = self.make_ty(&field.type_ref);
+            self.infer_pat(subpat, &Expectation::has_type(sub_ty));
+        }
+
+        ty
+    }
+
+    fn infer_struct(&mut self, path: Option<&Path>, subpats: &[FieldPat]) -> Ty {
+        let (ty, fields) = if let Some(x) = self.resolve_fields(path) {
+            x
+        } else {
+            return Ty::Unknown;
+        };
+
+        for subpat in subpats {
+            let matching_field = fields.iter().find(|field| field.name == subpat.name);
+
+            if let Some(field) = matching_field {
+                let typeref = &field.type_ref;
+                let sub_ty = self.make_ty(typeref);
+                self.infer_pat(subpat.pat, &Expectation::has_type(sub_ty));
+            }
+        }
+
+        ty
+    }
+
+    // TODO: Expectation should probably contain a Cow pointer to Ty?
+    // so that we can make new expectations of subtypes cheaply
+    fn infer_pat(&mut self, pat: PatId, expected: &Expectation) -> Ty {
+        let body = Arc::clone(&self.body); // avoid borrow checker problem
+
+        let ty = match &body[pat] {
+            Pat::Tuple(ref args) => {
+                // this can probably be done without cloning/ collecting
+                let expectations = match expected.ty {
+                    Ty::Tuple(ref tuple_args) if args.len() == tuple_args.len() => {
+                        tuple_args.iter().cloned().collect()
+                    }
+                    _ => vec![Ty::Unknown; args.len()],
+                };
+
+                let inner_tys = args
+                    .iter()
+                    .zip(expectations.into_iter())
+                    .map(|(&pat, ty)| self.infer_pat(pat, &Expectation::has_type(ty)))
+                    .collect::<Vec<_>>()
+                    .into();
+
+                Ty::Tuple(inner_tys)
+            }
+            Pat::Ref { pat, mutability } => {
+                let expectation = match expected.ty {
+                    Ty::Ref(ref sub_ty, exp_mut) if *mutability == exp_mut => {
```

Comment:
`*mutability != exp_mut` will be a type error, but it might be good to keep going anyway?

Reply:
I think in this particular case we actually keep going, just without any expectation ;-)